Mathlib/Tactic/Variable.lean

```diff
@@ -0,0 +1,299 @@
+`variable [Semiring R] [AddCommMonoid M] [Module R M]`, though if any of these three instance
+arguments can be inferred from previous variables then they will be omitted.
+
+An inherent limitation with this command is that variables are recorded in the scope as
+*syntax*. This means that `variable?` needs to pretty print the expressions we get
+from typeclass synthesis errors, and these might fail to round trip.
+-/
+
+namespace Mathlib.Command.Variable
+open Lean Elab Command Parser.Term Meta
+
+initialize registerTraceClass `variable?
+
+register_option variable?.maxSteps : Nat :=
+  { defValue := 15
+    group := "variable?"
+    descr :=
+      "The maximum number of instance arguments `variable?` will try to insert before giving up" }
+
+register_option variable?.checkRedundant : Bool :=
+  { defValue := true
+    group := "variable?"
+    descr := "Warn if instance arguments can be inferred from preceding ones" }
+
+/-- Get the type out of a bracketed binder. -/
+def bracketedBinderType : Syntax → Option Term
+  | `(bracketedBinderF|($_* $[: $ty?]? $(_annot?)?)) => ty?
+  | `(bracketedBinderF|{$_* $[: $ty?]?})             => ty?
+  | `(bracketedBinderF|⦃$_* $[: $ty?]?⦄)             => ty?
+  | `(bracketedBinderF|[$[$_ :]? $ty])               => some ty
+  | _                                                => none
+
+/-- The `variable?` command has the same syntax as `variable`, but it will auto-insert
+missing instance arguments wherever they are needed.
+It does not add variables that can already be deduced from others in the current context.
+By default the command checks that variables aren't implied by earlier ones, but it does *not*
+check that earlier variables aren't implied by later ones.
+Unlike `variable`, the `variable?` command does not support changing variable binder types.
+
+The `variable?` command will give a suggestion to replace itself with a command of the form
+`variable? ...binders... => ...binders...`.  The binders after the `=>` are the completed
+list of binders. When this `=>` clause is present, the command verifies that the expanded
+binders match the post-`=>` binders.  The purpose of this is to help keep code that uses
+`variable?` resiliant against changes to the typeclass hierarchy, at least in the sense
+that this additional information can be used to debug issues that might arise.
+One can also replace `variable? ...binders... =>` with `variable`.
+
+The core algorithm is to try elaborating binders one at a time, and whenever there is a
+typeclass instance inference failure, it synthesizes binder syntax for it and adds it to
+the list of binders and tries again, recursively. There are no guarantees that this
+process gives the "correct" list of binders.
+
+Structures tagged with the `variable_alias` attribute can serve as aliases for a collection
+of typeclasses. For example, given
+```lean
+@[variable_alias]
+structure VectorSpace (k V : Type _) [Field k] [AddCommGroup V] [Module k V]
+```
```

Comment:
Perhaps add explicit clarification that this definition should not have a body (i.e. fields of the structure)?